samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -150,6 +152,34 @@
       Serde<K> keySerde, Serde<M> messageSerde, Serde<OM> otherMessageSerde,
       Duration ttl, String id);
 
+  /**
+   * Joins this {@link MessageStream} with another {@link Table} using the provided
+   * pairwise {@link StreamTableJoinFunction}.
+   * <p>
+   * Messages are looked up from the joined table, join function is applied and join results are
+   * emitted as matches are found.
+   * <p>
+   * The join function allows implementation of both inner and left outer join. A null will be
```

Comment:
Agreed, it will be cleaner to separate them at API level. As it stands now, the join will support both through the join function. We can/should split the current join into inner and outer across the board for overall consistency.